Add simple talk/listener tutorial, matching the official ROS 2 documentation #378 (#390)
ros2-rust/ros2_rust: docs/writing_a_simple_publisher_and_subscriber.md

# Writing a simple publisher and subscriber (RUST)
* Goal: Create and run a publisher and subscriber node using Python.
* Tutorial level: Beginner
* Time: 20 minutes
<details><summary>Background</summary>

> In this tutorial you will create a 

Reviewer: ```suggestion
In this tutorial you will create a pair of
```

Author: done

[nodes](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Nodes/Understanding-ROS2-Nodes.html) that pass information to each other via a 
[topic](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Topics/Understanding-ROS2-Topics.html) in the form of string messages. The example used here is a simple "talker" and "listener" system; one node publishes data and the other subscribes to the topic to receive that data.

Since Rust doesn't have inheritance, it's not possible to inherit from `Node` as is common practice in [`rclcpp`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html) or [`rclpy`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html).

The code used in these examples can be found [here](https://gitlab.com/ros21923912/simple_ros2_node)  
<div style="margin-left:20px;">

> <details><summary>Sidenode to dependencies</summary>

Author: Has been inserted.

Almost none of the ROS2 dependencies you'll need for your ros2 rust node development currently exist on [crates.io](https://crates.io/), the main source for rust depencies. So the add command simply can't find the dependency targets. What colcon does by compiling the ros2 rust dependencies and your ros2 rust project is redirect the cargo search for dependencies directly into your `workspace/install` folder, where it'll find locally generated rust projects to use as dependencies. In particular, almost all message types will be called as dependencies for your ros2 rust project this way.

</details></div>

</details>

<details><summary>Prerequisites </summary> 

In previous tutorials, you learned how to create a [workspace](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-A-Workspace/Creating-A-Workspace.html) and [create a package](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-Your-First-ROS2-Package.html).

A basic understanding of [RUST](https://www.rust-lang.org/) is recommended, but not entirely necessary.
Before developing ROS2 RUST nodes, you must follow the 

> [installation instructions](https://github.com/ros2-rust/ros2_rust/blob/main/README.md) for them.

Author: done

</details>

<details><summary>Tasks </summary> 
<div style="margin-left:20px;"><details><summary>Create a Package</summary>

Currently, building a package for ROS2 RUST is different 
from building packages for Python or C/C++.  

First, you'll need to create and go into a standard [cargo](https://doc.rust-lang.org/cargo/) 
project as follows:
```
cargo new your_project_name && your_project_name
```
In the [`Cargo.toml`](https://doc.rust-lang.org/book/ch01-03-hello-cargo.html) file, add a dependency on `rclrs = "*"` and `std_msgs = "*"` by editing this file. For a full Introduction into RUST, please read the very good [RUST book](https://doc.rust-lang.org/book/title-page.html). Your `Cargo.toml` could now look like this:
```
[package]
name = "your_package_name"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rclrs = "*"
std_msgs = "*"
```

> Additionally, create a new `package.xml` if you want your node to be buildable with [`colcon`](https://colcon.readthedocs.io/en/released/user/installation.html). Make sure to change the build type to `ament_cargo` and to include the two packages mentioned above in the dependencies, as such:

Author: addressed

```xml
<package format="3">
  <name>your_project_name</name>
  <version>0.0.0</version>
  <description>TODO: Package description. Seriously. Please make a Package description. We all will thank for it.</description>
  <maintainer email="user@todo.todo">user</maintainer>
  <license>TODO: License declaration. Licenses are Great. Please add a Licence</license>

  <depend>rclrs</depend>
  <depend>std_msgs</depend>

  <export>
    <build_type>ament_cargo</build_type>
  </export>
</package>
```
By taking a look at your package, for example by typing [`tree .`](https://www.geeksforgeeks.org/tree-command-unixlinux/) inside your package, and you'll see a structure similar to the following:  
```
├── Cargo.toml
├── package.xml
└── src
    └── main.rs

2 directories, 3 files
```
Of course, you can use any capable editor or even your file explorer to do this.  

</details>

<details><summary>Write the publisher node</summary>

> To construct a node, replace the code in your main.rs with the [following](https://gitlab.com/ros21923912/simple_ros2_node/-/raw/more_simple_nodes/src/simple_publisher.rs?ref_type=heads):  

Author: done

```
/// Creates a SimplePublisherNode, initializes a node and publisher, and provides
/// methods to publish a simple "Hello World" message on a loop in separate threads.

/// Imports the Arc type from std::sync, used for thread-safe reference counting pointers,
/// and the StringMsg message type from std_msgs for publishing string messages.
use std::{sync::Arc,time::Duration,iter,thread};
use rclrs::{RclrsError,QOS_PROFILE_DEFAULT,Context,create_node,Node,Publisher};
use std_msgs::msg::String as StringMsg;
// / SimplePublisherNode struct contains node and publisher members.
// / Used to initialize a ROS 2 node and publisher, and publish messages.
struct SimplePublisherNode {
    node: Arc<Node>,
    _publisher: Arc<Publisher<StringMsg>>,
}

> /// Creates a new SimplePublisherNode by initializing a node and publisher.

Author: done

    /// Creates a StringMsg with "Hello World" as the data, publishes it on
    /// the `_publisher`, and returns a Result. This allows regularly publishing
    /// a simple message on a loop.
    fn publish_data(&self,inkrement:i32) -> Result<i32,RclrsError> {

        let msg: StringMsg = StringMsg {
            data: format!("Hello World {}",inkrement),
        };
        self._publisher.publish(msg).unwrap();
        Ok(inkrement+1_i32)
    }
}

/// The main function initializes a ROS 2 context, node and publisher,
/// spawns a thread to publish messages repeatedly, and spins the node
/// to receive callbacks.
/// 
/// It creates a context, initializes a SimplePublisherNode which creates
/// a node and publisher, clones the publisher to pass to the thread,  
/// spawns a thread to publish "Hello World" messages repeatedly, and
/// calls spin() on the node to receive callbacks. This allows publishing
/// messages asynchronously while spinning the node.
fn main() -> Result<(),RclrsError> {
    let context = Context::new(std::env::args()).unwrap();
    let publisher = Arc::new(SimplePublisherNode::new(&context).unwrap());
    let publisher_other_thread = Arc::clone(&publisher);

>     let mut iterator: i32=0;

Reviewer: "iterator" has a specific meaning within Rust. Perhaps "count" might be a better variable name?

Author: done

    thread::spawn(move || -> () {
        iter::repeat(()).for_each(|()| {
            thread::sleep(Duration::from_millis(1000));
            iterator=publisher_other_thread.publish_data(iterator).unwrap();
        });
    });
    rclrs::spin(publisher.node.clone())
}
```

> <details><summary>Examine the Code:</summary>

Reviewer: ```suggestion
<details><summary>Examining the code in detail:</summary>
```

Author: done

> #### This first 3 lines of the Rust code imports tools for thread synchronization, time handling, iteration, threading, ROS 2 communication, and string message publishing. It's likely setting up a ROS 2 node that publishes string messages.

Reviewer: ```suggestion
#### The first 3 lines of the Rust code imports tools for thread synchronization, time handling, iteration, threading, ROS 2 communication, and string message publishing.
```

Author: done

```
use std::{sync::Arc,time::Duration,iter,thread};
use rclrs::{RclrsError,QOS_PROFILE_DEFAULT,Context,create_node,Node,Publisher};
use std_msgs::msg::String as StringMsg;
```
* use std::{sync::Arc, time::Duration, iter, thread};: - Imports specific features from the standard library: - Arc is for thread-safe shared ownership of data. - Duration represents a time span. - iter provides tools for working with iterators. - thread enables creating and managing threads.

> * use rclrs::{RclrsError, QOS_PROFILE_DEFAULT, Context, create_node, Node, Publisher};: - Imports elements for ROS 2 communication: - RclrsError for handling errors. - QOS_PROFILE_DEFAULT likely for default Quality of Service settings. - Context, create_node, Node, Publisher are for ROS 2 node creation and publishing.

Reviewer: ```suggestion
* use rclrs::{RclrsError, QOS_PROFILE_DEFAULT, Context, create_node, Node, Publisher};: - Imports elements for ROS 2 communication: - RclrsError for handling errors. - QOS_PROFILE_DEFAULT for default Quality of Service settings. - Context, create_node, Node, Publisher are for ROS 2 node creation and publishing.
```

Author: done

* use std_msgs::msg::String as StringMsg;: - Imports the StringMsg type for publishing string messages.  

> #### Next this struct defines a SimplePublisherNode which holds references to a ROS 2 node and a publisher for string messages.

Reviewer: ```suggestion
#### Next, this structure defines a SimplePublisherNode which holds references to a ROS 2 node and a publisher for string messages.
```

Author: done

2. Members:
* node: Arc<Node>: This member stores a reference to a ROS 2 node, wrapped in an Arc (Atomic Reference Counted) smart pointer. This allows for safe sharing of the node reference across multiple threads.  
* _publisher: Arc<Publisher<StringMsg>>: This member stores a reference to a publisher specifically for string messages (StringMsg), also wrapped in an Arc for thread safety. The publisher is responsible for sending string messages to other nodes in the ROS 2 system.  

3. This code defines methods for the SimplePublisherNode struct. The new method creates a ROS 2 node and publisher, storing them in the struct. The publish_data method publishes a string message with a counter and returns the incremented counter.
```
impl SimplePublisherNode {
    fn new(context: &Context) -> Result<Self,RclrsError> {
        let node = create_node(context, "simple_publisher").unwrap();
        let _publisher = node
            .create_publisher("publish_hello", QOS_PROFILE_DEFAULT)
            .unwrap();
        Ok(Self { node, _publisher, })
    }

>     fn publish_data(&self,inkrement:i32) -> Result<i32,RclrsError> {

Reviewer: ```suggestion
    fn publish_data(&self, increment:i32) -> Result<i32, RclrsError> {
```

Author: done

        let msg: StringMsg = StringMsg {

>             data: format!("Hello World {}",inkrement),

Reviewer: ```suggestion
            data: format!("Hello World {}", increment),
```

Author: done

        };
        self._publisher.publish(msg).unwrap();

>         Ok(inkrement+1_i32)

Reviewer: ```suggestion
        Ok(increment + 1_i32)
```
The `i32` type information may not be needed - Rust's compiler is pretty good about figuring out the types of things.

Author: done

    }
}
```

1. Implementation Block:
`impl SimplePublisherNode { ... }`: This line indicates that methods are being defined for the `SimplePublisherNode` struct.

2. Constructor Method:
* `fn new(context: &Context) -> Result<Self, RclrsError> { ... }`: This method serves as a constructor for creating instances of SimplePublisherNode.
    * It takes a Context object as input, which is necessary for interacting with the ROS 2 system.
    * It returns a Result type, indicating either a successful Self (the created SimplePublisherNode object) or an RclrsError if something goes wrong.
    * Inside the new method:

>         * `let node = create_node(context, "simple_publisher").unwrap();`: Creates a new ROS 2 node named "simple_publisher" within the given context. The unwrap() unwraps the result, handling any errors immediately.

Reviewer: ```suggestion
        * `let node = create_node(context, "simple_publisher").unwrap();`: Creates a new ROS 2 node named "simple_publisher" within the given context. The unwrap() unwraps the result, handling any errors immediately by forcing the program to abort (`panic`) if something goes wrong. Since our code can't function properly if the node is not able to be created, this is a valid error-handling response for our use-case.
```
Just adding a little note on `unwrap` here - it's a function that trips up a lot of new users, so I like to have tutorials that let them know what it's going to do. Some people may be a bit surprised that their program crashes when it's used!

Author: done

        * `let _publisher = node.create_publisher("publish_hello", QOS_PROFILE_DEFAULT).unwrap();`: Creates a publisher for string messages on the topic "publish_hello" with default quality of service settings.

>         * `Ok(Self { node, _publisher, })`: Returns a Result with the newly created SimplePublisherNode object, containing the node and publisher references.

Reviewer: ```suggestion
        * `Ok(Self { node, _publisher, })`: Returns an `Ok` Result with the newly created SimplePublisherNode object, containing the node and publisher references.
```

Author: done

3. Publishing Method:

> * `fn publish_data(&self, inkrement: i32) -> Result<i32, RclrsError> { ... }`: This method publishes a string message and increments a counter.

Reviewer: ```suggestion
* `fn publish_data(&self, increment: i32) -> Result<i32, RclrsError> { ... }`: This method publishes a string message and increments a counter.
```

Author: done

>     * It takes an inkrement value (an integer) as input, which is likely used for counting purposes within the message content.

Reviewer: ```suggestion
    * It takes an increment value (an integer) as input, which is used for counting purposes within the message content.
```

Author: done

>     * It also returns a Result type, indicating either the incremented inkrement value or an RclrsError if publishing fails.

Reviewer: ```suggestion
    * It also returns a Result type, indicating either the incremented value or an RclrsError if publishing fails.
```

Author: done

    * Inside the publish_data method:
        * `let msg: StringMsg = StringMsg { data: format!("Hello World {}", inkrement), };`:tCreates a string message with the content "Hello World" followed by the inkrement value.
        * self._publisher.publish(msg).unwrap();: Publishes the created message onto the topic associated with the publisher.

>         * Ok(inkrement + 1_i32): Returns a Result with the incremented inkrement value.  

Author: done

#### The main Method creates a ROS 2 node that publishes string messages at a rate of 1 Hz.

```
fn main() -> Result<(),RclrsError> {
   let context = Context::new(std::env::args()).unwrap();
   let publisher = Arc::new(SimplePublisherNode::new(&context).unwrap());
   let publisher_other_thread = Arc::clone(&publisher);
   let mut iterator: i32=0;
   thread::spawn(move || -> () {
       iter::repeat(()).for_each(|()| {
           thread::sleep(Duration::from_millis(1000));

>            iterator=publisher_other_thread.publish_data(iterator).unwrap();

Reviewer: Again, I feel like "count" or "counter" might be a better variable name than "iterator" here. Especially because we're already using an iterator on line 248.

Author: done

* `thread::spawn(move || -> () { ... });`: Spawns a new thread with a closure: `iter::repeat(()).for_each(|()| { ... });`: Creates an infinite loop using `iter::repeat`.

4. Publishing Loop within Thread:

* `thread::sleep(Duration::from_millis(1000));`: Pauses the thread for 1 second (1 Hz publishing rate).
* `iterator = publisher_other_thread.publish_data(iterator).unwrap();`: Calls the publish_data method on the publisher_other_thread to publish a message with the current counter value. Increments the iterator for the next message.

5. Main Thread Spin:
* `rclrs::spin(publisher.node.clone());`: Keeps the main thread running, processing ROS 2 events and messages. Uses a cloned reference to the node to ensure it remains active even with other threads.

</details>

Once you have implemented the code, you are ready to run it:
```
cd ${MainFolderOfWorkspace}
colcon build
source install/setub.bash
```
And finally run with:
```
ros2 run your_project_name your_project_name
```
(Please give your package a better name than me ;) )
</details>
<details><summary>Having several Ros2 rust nodes in one Package</summary>

> Of course, you can write for each node you want to implement its own package, and that can have it's advantages. I implore you to use some cargo tricks and add some binary targets to your `cargo.toml`. this could look like this:

Reviewer: ```suggestion
Of course, you can write for each node you want to implement its own package, and that can have it's advantages. I implore you to use some cargo tricks and add some binary targets to your `cargo.toml`. That could look like this:
```

Author: done

```
[package]
name = "your_package_name"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[[bin]]
name="simple_publisher"
path="src/main.rs"
[dependencies]
rclrs = "*"
std_msgs = "*"
```
You'll find the name of your executable and the corresponding file name under the `[[bin]]` tag. As you can see, the filename and the name you want to call your node don't have to match. Please remember to include your executable name with snake_cases. The rust compiler will be a bit grumpy if you don't.  
Now, by recompiling the package from the previous chapter and making it usable:  
```
cd ${MainFolderOfWorkspace}
colcon build
source install/setub.bash
```

> node will look like this:

Reviewer: ```suggestion
Running the node will look like this:
```

Author: done

```
ros2 run your_package_name simple_publisher
```
As you can see, you are now calling your node by the name declared in `[[bin]]` using the `name` variable.
</details>
<details><summary>Write the subscriber node</summary> 

Of course, you can implement a new ros2 rust package for this node. You can find out how to do this in the section called 'Create a package'.

> Or you can add a new binary target to your package. Then just add a new `<file>.rs` to your source directory - for simplicity I'll call this file `simple_subscriber.rs` - and add a corresponding binary target to your `Cargo.toml`:

Reviewer: ```suggestion
Or you can add a new binary target to your package. To do so, just add a new `<file>.rs` to your source directory - for simplicity I'll call this file `simple_subscriber.rs` - and add a corresponding binary target to your `Cargo.toml`:
```

Author: done

# Writing a simple publisher and subscriber (Rust)
* Goal: Create and run a publisher and subscriber node using Rust.
* Tutorial level: Beginner
* Time: 20 minutes
<details><summary>Background</summary>

In this tutorial you will create a pair of 
[nodes](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Nodes/Understanding-ROS2-Nodes.html) that pass information to each other via a 
[topic](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Topics/Understanding-ROS2-Topics.html) in the form of string messages. The example used here is a simple "talker" and "listener" system; one node publishes data and the other subscribes to the topic to receive that data.

Since Rust doesn't have inheritance, it's not possible to inherit from `Node` as is common practice in [`rclcpp`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html) or [`rclpy`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html).

> The code used in these examples can be found [here](https://gitlab.com/ros21923912/simple_ros2_node/-/tree/more_simple_nodes?ref_type=heads)  

Reviewer 2: Can you explain in detail what are the issues you're having building `ros2_rust` locally? I really would want the examples to be in a folder in this PR (named `examples`)

Author: The main problem is that clippy just won't start on my computer because it doesn't understand how the ros2 installation works with rust.

* [packages](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-Your-First-ROS2-Package.html).

A basic understanding of [Rust](https://doc.rust-lang.org/book/) is recommended, but not entirely necessary.
Before developing [ros2-rust](https://github.com/ros2-rust/ros2_rust) nodes, you must follow the 
[installation instructions](https://github.com/ros2-rust/ros2-rust/blob/main/README.md) for [`rclrs`](https://docs.rs/rclrs/latest/rclrs/).
For a full Introduction into Rust, please read the very good [Rust book](https://doc.rust-lang.org/book/title-page.html).

</details>

<details><summary>Tasks </summary> 
<div style="margin-left:20px;"><details><summary>Create a Package</summary>

Currently, building a package for ros2-rust is different 
from building packages for [Python](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html) or [C/C++](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html).  

First, you'll need to create and go into a standard [cargo](https://doc.rust-lang.org/cargo/) 
project as follows:
```

> cargo new rust_pub_sub && cd rust_pub_sub

Reviewer 2: Like @maspe36 suggested and also in order to follow the C++ examples, please name this `rust_pubsub` (like in C++ it's called `cpp_pubsub`)

##### new
The `new` method creates a ROS 2 node, subscriber, and a storage location for received messages, storing them in the `struct`.
```rust
    fn new(context: &Context) -> Result<Self, RclrsError> {
        let node = create_node(context, "simple_subscription").unwrap();
        let data: Arc<Mutex<Option<StringMsg>>> = Arc::new(Mutex::new(None));
        let data_mut: Arc<Mutex<Option<StringMsg>>> = Arc::clone(&data);
        let _subscriber = node
            .create_subscription::<StringMsg, _>(
                "publish_hello",
                QOS_PROFILE_DEFAULT,
                move |msg: StringMsg| {
                    *data_mut.lock().unwrap() = Some(msg);
                },
            )
            .unwrap();
        Ok(Self {
            node,
            _subscriber,
            data,
        })
    }

```
A few special features:  
1. Initializing Shared Data:  
    * `let data: Arc<Mutex<Option<StringMsg>>> = Arc::new(Mutex::new(None));`  
        This line creates a shared data structure that will hold the received message.  
        * `Arc<Mutex<Option<StringMsg>>>`: This is a complex type combining several functionalities:  

>             * `Arc<T>`: An atomically reference-counted pointer ([`Arc`](https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://doc.rust-lang.org/std/sync/struct.Arc.html&ved=2ahUKEwiJz_n3876FAxXZg_0HHc-yDZ8QFnoECAYQAQ&usg=AOvVaw2ZAPxD2olFejU3a_Ngb4f5)) allows multiple parts of the code to safely access the same data (`T`).

Reviewer 2: Please use the real links from https://doc.rust-lang.org instead of Google links.

Author: done

```sh
cd WORKSPACE
source install/setup.bash
ros2 run rust_pub_sub your_node_name
```
In my case, the nodes are called `simple_publisher` and `simple_subscriber`. You can name your nodes whatever you like. It is important that the publisher and subscriber use the same topic type and name.  
If you haven't had any errors so far and have successfully started the Publisher and Subscriber, you should see something similar in the Subscriber's Terminal window:
```sh
Hello World 230
Hello World 231
Hello World 232
Hello World 233
Hello World 234
Hello World 235
Hello World 236
Hello World 237
Hello World 238
Hello World 239
Hello World 240
Hello World 241
Hello World 242
Hello World 243
Hello World 244
Hello World 245
Hello World 246
```
(My nodes have been running for some time.)  
Enter `Ctrl+c` in each terminal to stop the nodes from spinning.
</details></div>
</details>

<details><summary>Summary</summary>

You created two nodes to publish and subscribe to data over a topic. Before running them, you added their dependencies and entry points to the package configuration files.

</details></details>

<details><summary>Last thoughts</summary>

> At the end of the day, tools must not only work more safely and efficiently from a purely rational point of view, but they must also give the end user, as well as the developer, a good time. Hopefully you had fun developing the two nodes. Without fun, software development can be boring and will often prevent you from using this tool again. 

Reviewer 2: I really like this, nice touch. No feedback, just wanted to highlight it :slightly_smiling_face:

Author: Thanks

# Writing a simple publisher and subscriber (Rust)
* Goal: Create and run a publisher and subscriber node using Rust.
* Tutorial level: Beginner
* Time: 20 minutes
<details><summary>Background</summary>

In this tutorial you will create a pair of 
[nodes](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Nodes/Understanding-ROS2-Nodes.html) that pass information to each other via a 
[topic](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Topics/Understanding-ROS2-Topics.html) in the form of string messages. The example used here is a simple "talker" and "listener" system; one node publishes data and the other subscribes to the topic to receive that data.

> Since Rust doesn't have inheritance, it's not possible to inherit from `Node` as is common practice in [`rclcpp`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html) or [`rclpy`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html).

Reviewer 3: I think this section would be more appropriate right before we instruct the user to define a node. Its out of place up here.

Author: To be honest: I don't see the point in rebuilding the whole text now, after so many changes have already been made. I just want to finalise or finish the whole thing. I've literally been sitting on this pull request for months and I just don't want any more.

The code used in these examples can be found [here](https://github.com/ros2-rust/ros2_rust/tree/main/examples/rust_pubsub)
<div style="margin-left:20px;">
<details><summary>Side-note on dependencies</summary>

> You may be wondering why you can't just add all your ROS 2-specific dependencies to `Cargo.toml` with `cargo add YOUR_DEPENDENCIES` and have to edit this file manually. Here is why:

Reviewer 3: IMO this is also out of place this far up. This is super important information to capture, but I think this would be more appropriate when we actually talk about adding dependencies to nodes

Author: I think this information is important for a newcomer, as a normal Rust user who has worked with the command line tools to some extent will be confused as to why this suddenly no longer works.

* [packages](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-Your-First-ROS2-Package.html).

A basic understanding of [Rust](https://doc.rust-lang.org/book/) is recommended, but not entirely necessary.
Before developing [ros2-rust](https://github.com/ros2-rust/ros2_rust) nodes, you must follow the 
[installation instructions](https://github.com/ros2-rust/ros2-rust/blob/main/README.md) for [`rclrs`](https://docs.rs/rclrs/latest/rclrs/).
For a full Introduction into Rust, please read the very good [Rust book](https://doc.rust-lang.org/book/title-page.html).

</details>

<details><summary>Tasks </summary> 
<div style="margin-left:20px;"><details><summary>Create a Package</summary>

Currently, building a package for ros2-rust is different 
from building packages for [Python](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html) or [C/C++](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html).  

First, you'll need to create a standard [cargo](https://doc.rust-lang.org/cargo/) 
package:

> ```

Reviewer 2: ```suggestion
```console
```

Author: addressed

Author: I didn't knew that.

Author: That was too inconsistent for me. I have replaced each `shell` and `console` hint with an `sh`. so it is consistent for normal terminal calls.

cargo new rust_pubsub && cd rust_pubsub
```
In the [`Cargo.toml`](https://doc.rust-lang.org/book/ch01-03-hello-cargo.html) file, add a dependency on `rclrs = "*"` and `std_msgs = "*"` by editing this file. Your `Cargo.toml` should now look like this:
```toml
[package]
name = "rust_pubsub"
version = "0.1.0"
edition = "2021"

> 
> # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
> 

Reviewer 3: ```suggestion

```

[dependencies]
rclrs = "*"
std_msgs = "*"
```

> 
> 

Reviewer 3: ```suggestion

```

    });
    rclrs::spin(publisher.node.clone())
}
```

<details><summary>Examining the code in detail:</summary>

#### Importing 
The first 3 lines of the Rust code imports tools for thread synchronization, time 
handling, iteration, threading, ROS 2 communication, and string message publishing.
```rust
use rclrs::{create_node, Context, Node, Publisher, RclrsError, QOS_PROFILE_DEFAULT};
use std::{env, sync::Arc, thread, time::Duration};
use std_msgs::msg::String as StringMsg;
```
* `use std::{sync::Arc, time::Duration, iter, thread};`: Imports specific features from the standard library: 
    - `Arc` is for thread-safe shared ownership of data. 
    - `Duration` represents a time span. 
    - `iter` provides tools for working with iterators.
    - `thread` enables creating and managing threads.
* `use rclrs::{RclrsError, QOS_PROFILE_DEFAULT, Context, create_node, Node, Publisher};`: 
    - Imports elements for ROS 2 communication: 
        - `RclrsError` for handling errors. 
        - `QOS_PROFILE_DEFAULT` for default Quality of Service settings. 
- `Context, create_node, Node, Publisher` are for ROS 2 node creation and publishing.

> * `use std_msgs::msg::String as StringMsg;`: Imports the `StringMsg` type for publishing string messages.  

Reviewer 3: The order of the imports in the example code does not match the order of the bullet points where we expand whats going on in each line.

* `let mut iterator: i32 = 0;`: Initializes a counter variable for message content.  
* `thread::spawn(move || -> () { ... });`: Spawns a new [thread](https://doc.rust-lang.org/std/thread/index.html) with a [closure](https://doc.rust-lang.org/book/ch13-01-closures.html): `loop { ... }`: Creates an infinite loop using [`loop`](https://doc.rust-lang.org/std/keyword.loop.html).  
4. Publishing Loop within Thread:  
* `thread::sleep(Duration::from_millis(1000));`: Pauses the thread for 1 second (1 Hz publishing rate).  
* `iterator = publisher_other_thread.publish_data(count).unwrap();`: Calls the `publish_data` method on the `publisher_other_thread` to publish a message with the current counter value. Increments the iterator for the next message.  
5. Main Thread Spin:  
* `rclrs::spin(publisher.node.clone());`: Keeps the main thread running, processing ROS 2 events and messages. Uses a cloned reference to the node to ensure it remains active even with other threads.  

</details>
</details>
<details><summary>Having several ROS 2 Rust nodes in one Package</summary>

> Of course, you can write for each node you want to implement its own package, and that can have it's advantages. I implore you to use some cargo tricks and add some binary targets to your `cargo.toml`. That could look like this:

Reviewer 3: The tutorial should be in the second (or third) person, not first person. This means avoiding wording like "I impore you" and to phrase it like "You may find it useful".

```toml
[package]
name = "rust_pubsub"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[[bin]]
name="simple_publisher"
path="src/simple_publisher.rs"
[dependencies]
rclrs = "*"
std_msgs = "*"
```

You'll find the name of your executable and the corresponding file name under the `[[bin]]` tag. As you can see, the filename and the name you want to call your node don't have to match. Please remember to include your executable name with snake_cases. The Rust compiler will be a bit grumpy if you don't.  
Now, by recompiling the package from the previous chapter and making it usable:  
```sh

> cd WORKSPACE

Reviewer 3: We should probably have them create a simple workspace as well as part of this tutorial. We're asking them to `cd` into something that may/may not exist. I know we say you should have knowledge of workspaces, but I think a small section in the prereqs could be good.

Author: It has already been mentioned elsewhere that knowledge of ros2 workspaces and how to set them up is a prerequisite for this tutorial.

# Writing a simple publisher and subscriber (Rust)
* Goal: Create and run a publisher and subscriber node using Rust.
* Tutorial level: Beginner
* Time: 20 minutes
<details><summary>Background</summary>

In this tutorial you will create a pair of 
[nodes](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Nodes/Understanding-ROS2-Nodes.html) that pass information to each other via a 
[topic](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Topics/Understanding-ROS2-Topics.html) in the form of string messages. The example used here is a simple "talker" and "listener" system; one node publishes data and the other subscribes to the topic to receive that data.

Since Rust doesn't have inheritance, it's not possible to inherit from `Node` as is common practice in [`rclcpp`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html) or [`rclpy`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html).

> The code used in these examples can be found [here](https://gitlab.com/ros21923912/simple_ros2_node/-/tree/more_simple_nodes?ref_type=heads)  

Reviewer 2: This needs to be updated to point to the example included in this PR

Author: I have now removed the link. That should be a better fit.

Reviewer 2: Just like this:

The code used in these examples can be found [here](https://github.com/ros2-rust/ros2_rust/tree/main/examples/rust_pubsub)

Doesn't matter if this does not exist now, it will exist when this is merged.

* [workspaces](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-A-Workspace/Creating-A-Workspace.html)
* [packages](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-Your-First-ROS2-Package.html).

A basic understanding of [Rust](https://doc.rust-lang.org/book/) is recommended, but not entirely necessary.
Before developing [ros2-rust](https://github.com/ros2-rust/ros2_rust) nodes, you must follow the 
[installation instructions](https://github.com/ros2-rust/ros2-rust/blob/main/README.md) for [`rclrs`](https://docs.rs/rclrs/latest/rclrs/).

</details>

<details><summary>Tasks </summary> 
<div style="margin-left:20px;"><details><summary>Create a Package</summary>

Currently, building a package for ros2-rust is different 
from building packages for [Python](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html) or [C/C++](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html).  

First, you'll need to create and go into a standard [cargo](https://doc.rust-lang.org/cargo/) 
project as follows:
```

> cargo new your_package_name && cd your_package_name

Reviewer 3: Lets pick a package name. Perhaps `rust_pubsub`? That sort of matches the official ros2 docs 

> ros2 pkg create --build-type ament_cmake cpp_pubsub

https://docs.ros.org/en/foxy/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html#create-a-package

```

> In the [`Cargo.toml`](https://doc.rust-lang.org/book/ch01-03-hello-cargo.html) file, add a dependency on `rclrs = "*"` and `std_msgs = "*"` by editing this file. For a full Introduction into Rust, please read the very good [Rust book](https://doc.rust-lang.org/book/title-page.html). Your `Cargo.toml` could now look like this:

Author: Good point. Is done

```toml
[package]
name = "your_package_name"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rclrs = "*"
std_msgs = "*"
```

Additionally, create a new `package.xml` if you want your node to be buildable with [`colcon`](https://colcon.readthedocs.io/en/released/user/installation.html). Make sure to change the build type to `ament_cargo` and to include the two packages mentioned above in the dependencies, as such:
```xml
<package format="3">
  <name>your_package_name</name>
  <version>0.0.0</version>
  <description>TODO: Package description.</description>
  <maintainer email="user@todo.todo">user</maintainer>
  <license>TODO: License declaration.</license>

  <depend>rclrs</depend>
  <depend>std_msgs</depend>

  <export>
    <build_type>ament_cargo</build_type>
  </export>
</package>
```

> By taking a look at your package, for example by typing [`tree .`](https://www.geeksforgeeks.org/tree-command-unixlinux/) inside your package, and you'll see a structure similar to the following:  

Author: done

```shell
├── Cargo.toml
├── package.xml
└── src
    └── main.rs

> 2 directories, 3 files

Author: done

```

> Of course, you can use any capable editor or even your file explorer to do this.  

Author: Is done

</details>

<details><summary>Write the publisher node</summary>

To construct a node, replace the code in your `main.rs` file with the [following](https://gitlab.com/ros21923912/simple_ros2_node/-/blob/more_simple_nodes/src/simple_publisher.rs?ref_type=heads):  
```rust
/// Creates a SimplePublisherNode, initializes a node and publisher, and provides
/// methods to publish a simple "Hello World" message on a loop in separate threads.

/// Imports the Arc type from std::sync, used for thread-safe reference counting pointers,

> /// and the StringMsg message type from std_msgs for publishing string messages.

Author: done

use std::{sync::Arc,time::Duration,iter,thread};
use rclrs::{RclrsError,QOS_PROFILE_DEFAULT,Context,create_node,Node,Publisher};
use std_msgs::msg::String as StringMsg;
// / SimplePublisherNode struct contains node and publisher members.
// / Used to initialize a ROS 2 node and publisher, and publish messages.
struct SimplePublisherNode {
    node: Arc<Node>,

>     _publisher: Arc<Publisher<StringMsg>>,

Author: done

        };
        self._publisher.publish(msg).unwrap();
        Ok(increment+1_i32)
    }
}

/// The main function initializes a ROS 2 context, node and publisher,
/// spawns a thread to publish messages repeatedly, and spins the node
/// to receive callbacks.
/// 
/// It creates a context, initializes a SimplePublisherNode which creates
/// a node and publisher, clones the publisher to pass to the thread,  
/// spawns a thread to publish "Hello World" messages repeatedly, and
/// calls spin() on the node to receive callbacks. This allows publishing
/// messages asynchronously while spinning the node.
fn main() -> Result<(),RclrsError> {
    let context = Context::new(std::env::args()).unwrap();
    let publisher = Arc::new(SimplePublisherNode::new(&context).unwrap());
    let publisher_other_thread = Arc::clone(&publisher);
    let mut count: i32=0;
    thread::spawn(move || -> () {
        iter::repeat(()).for_each(|()| {
            thread::sleep(Duration::from_millis(1000));
            count=publisher_other_thread.publish_data(count).unwrap();
        });
    });
    rclrs::spin(publisher.node.clone())
}
```

<details><summary>Examining the code in detail:</summary>

> #### The first 3 lines of the Rust code imports tools for thread synchronization, time handling, iteration, threading, ROS 2 communication, and string message publishing.

Reviewer 3: ```suggestion
#### Imports
The first 3 lines of the Rust code imports tools for thread synchronization, time handling, iteration, threading, ROS 2 communication, and string message publishing.
```

Author: done

```rust
use std::{sync::Arc,time::Duration,iter,thread};
use rclrs::{RclrsError,QOS_PROFILE_DEFAULT,Context,create_node,Node,Publisher};
use std_msgs::msg::String as StringMsg;
```
* `use std::{sync::Arc, time::Duration, iter, thread};`: Imports specific features from the standard library: 
    - `Arc` is for thread-safe shared ownership of data. 
    - `Duration` represents a time span. 

>     - `iter` provides tools for working with iterators. - thread enables creating and managing threads.

Author: Changes have been incorporated.

* `use rclrs::{RclrsError, QOS_PROFILE_DEFAULT, Context, create_node, Node, Publisher};`: 
    - Imports elements for ROS 2 communication: 
        - `RclrsError` for handling errors. 
        - `QOS_PROFILE_DEFAULT` for default Quality of Service settings. 
        - `Context, create_node, Node, Publisher` are for ROS 2 node creation and publishing. publishing.
* `use std_msgs::msg::String as StringMsg;`: Imports the `StringMsg` type for publishing string messages.  

> #### Next, this structure defines a SimplePublisherNode which holds references to a ROS 2 node and a publisher for string messages.

Reviewer 3: This header is pretty long, maybe this is easier for a reader

```suggestion
#### SimplePublisherNode
Next, this structure defines a SimplePublisherNode which holds references to a ROS 2 node and a publisher for string messages.
```

```rust
struct SimplePublisherNode {
    node: Arc<Node>,
    _publisher: Arc<Publisher<StringMsg>>,
}
```
1. Structure:  
`struct SimplePublisherNode`: This line defines a new [`struct`](https://doc.rust-lang.org/rust-by-example/custom_types/structs.html) named `SimplePublisherNode`. It serves as a blueprint for creating objects that hold information related to a simple publisher node in ROS 2.  

2. Members:
* `node: Arc<Node>`: This member stores a reference to a ROS 2 node, wrapped in an [`Arc` (Atomic Reference Counted)](https://doc.rust-lang.org/std/sync/struct.Arc.html) smart pointer. This allows for safe sharing of the node reference across multiple threads.  
* `_publisher: Arc<Publisher<StringMsg>>`: This member stores a reference to a publisher specifically for string messages (`StringMsg`), also wrapped in an `Arc` for thread safety. The publisher is responsible for sending string messages to other nodes in the ROS 2 system.  

> #### This code defines methods for the `SimplePublisherNode` `struct`. The `new` method creates a ROS 2 node and publisher, storing them in the `struct`. The `publish_data` method publishes a string message with a `counter` and returns the incremented `counter`.

Reviewer 3: ```suggestion
#### impl SimplePublisher
This code defines methods for the `SimplePublisherNode` `struct`. The `new` method creates a ROS 2 node and publisher, storing them in the `struct`. The `publish_data` method publishes a string message with a `counter` and returns the incremented `counter`.
```

    fn new(context: &Context) -> Result<Self,RclrsError> {
        let node = create_node(context, "simple_publisher").unwrap();
        let _publisher = node
            .create_publisher("publish_hello", QOS_PROFILE_DEFAULT)
            .unwrap();
        Ok(Self { node, _publisher, })
    }
    fn publish_data(&self,increment:i32) -> Result<i32,RclrsError> {

        let msg: StringMsg = StringMsg {
            data: format!("Hello World {}",increment),
        };
        self._publisher.publish(msg).unwrap();
        Ok(increment+1_i32)
    }
}
```

1. Implementation Block:   
`impl SimplePublisherNode { ... }`: This line indicates that methods are being defined for the `SimplePublisherNode` struct.  
2. Constructor Method:  
* `fn new(context: &Context) -> Result<Self, RclrsError> { ... }`: This method serves as a constructor for creating instances of SimplePublisherNode.  
    * It takes a Context object as input, which is necessary for interacting with the ROS 2 syste.  
    * It returns a Result type, indicating either a successful Self (the created `SimplePublisherNode` object) or an `RclrsError` if something goes wrong.  
    * Inside the new method:  

>         * `let node = create_node(context, "simple_publisher").unwrap();`: Creates a new ROS 2 node named `"simple_publisher"` within the given context. The [`unwrap()`](https://doc.rust-lang.org/rust-by-example/error/option_unwrap.html) unwraps the [`Result`](https://doc.rust-lang.org/std/result/), handling any errors immediately by forcing the program to abort (`panic`) if something goes wrong. Since our code can't function properly if the node is not able to be created, this is a valid error-handling response for our use-case.  

Reviewer 3: Since we're already linking to a lot of rust docs here, may as well link to panic as well

```suggestion
        * `let node = create_node(context, "simple_publisher").unwrap();`: Creates a new ROS 2 node named `"simple_publisher"` within the given context. The [`unwrap()`](https://doc.rust-lang.org/rust-by-example/error/option_unwrap.html) unwraps the [`Result`](https://doc.rust-lang.org/std/result/), handling any errors immediately by forcing the program to abort ([`panic`](https://doc.rust-lang.org/book/ch09-01-unrecoverable-errors-with-panic.html)) if something goes wrong. Since our code can't function properly if the node is not able to be created, this is a valid error-handling response for our use-case.
```

        * `Ok(Self { node, _publisher, })`: Returns an `Ok` Result with the newly created `SimplePublisherNode` object, containing the node and publisher references.  
3. Publishing Method:
* `fn publish_data(&self, increment: i32) -> Result<i32, RclrsError> { ... }`: This method publishes a string message and increments a `counter`.
    * It takes an inkrement value (an integer) as input, which is used for counting purposes within the message content.
    * It also returns a Result type, indicating either the incremented value or an RclrsError if publishing fails.
    * Inside the publish_data method:
        * `let msg: StringMsg = StringMsg { data: format!("Hello World {}", increment), };`: Creates a string message with the content `"Hello World"` followed by the increment value.
        * `self._publisher.publish(msg).unwrap();`: Publishes the created message onto the topic associated with the publisher.
        * `Ok(increment + 1_i32)`: Returns a Result with the incremented increment value.  

> #### The main Method creates a ROS 2 node that publishes string messages at a rate of 1 Hz.  

Reviewer 3: ```suggestion
#### main
The main Method creates a ROS 2 node that publishes string messages at a rate of 1 Hz.
```

```
Running the node will look like this:
```shell
ros2 run your_package_name simple_publisher
```
As you can see, you are now calling your node by the name declared in `[[bin]]` using the `name` variable.

</details>
<details><summary>Write the subscriber node</summary> 

Of course, you can implement a new ROS2 Rust package for this node. You can find out how to do this in the section called 'Create a package'.
Or you can add a new binary target to your package. To do so, just add a new `FILE.rs` to your source directory - for simplicity I'll call this file `simple_subscriber.rs` - and add a corresponding binary target to your `Cargo.toml`:
```toml
[[bin]]
name="simple_subscriber"
path="src/simple_subscriber.rs"
```
To construct the subscriber node, put the [following](https://gitlab.com/ros21923912/simple_ros2_node/-/blob/more_simple_nodes/src/simple_subscriber.rs?ref_type=heads) code into a `FILE.rs` - in my case its the `src/simple_subscriber.rs`:
```rust
use rclrs::{create_node, Context, Node, RclrsError, Subscription, QOS_PROFILE_DEFAULT};
use std::{
    env,
    iter,thread,
    sync::{Arc, Mutex},
    time::Duration,
};
use std_msgs::msg::String as StringMsg;

> /// A simple ROS2 subscriber node that receives and prints "hello" messages.

Reviewer 3: Because there is a section examining the code in detail, we should probably remove the more in depth descriptions from the comments. This is what the official ROS 2 docs do
https://docs.ros.org/en/foxy/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html#write-the-publisher-node

Notice that there is only a single comment in their actual example code
```
/* This example creates a subclass of Node and uses std::bind() to register a
* member function as a callback from the timer. */
```

Reviewer 2: @Guelakais please address this

Author: I have now deleted a few comments. That should clear that up too.

        } else {
            println!("No message available yet.");
        }
        Ok(())
    }
}
/// The `main` function creates a new ROS2 context, a `SimpleSubscriptionNode` instance, and starts a separate thread to periodically call the `data_callback` method on the subscription. The main thread then calls `rclrs::spin()` to keep the node running and receive messages.
///
/// The separate thread is used to ensure that the `data_callback` method is called regularly, even if the main thread is blocked in `rclrs::spin()`. This allows the subscriber to continuously process and print the received "hello" messages.
fn main() -> Result<(), RclrsError> {
    let context = Context::new(env::args()).unwrap();
    let subscription = Arc::new(SimpleSubscriptionNode::new(&context).unwrap());
    let subscription_other_thread = Arc::clone(&subscription);
    thread::spawn(move || -> () {
        iter::repeat(()).for_each(|()| {
            thread::sleep(Duration::from_millis(1000));
            subscription_other_thread.data_callback().unwrap()
        });
    });
    rclrs::spin(subscription.node.clone())
}
```
<details><summary>Examining the code in detail:</summary>

> #### The main Construct:

Reviewer 3: ```suggestion
#### SimpleSubscriptionNode
```

```rust
pub struct SimpleSubscriptionNode {
    node: Arc<Node>,
    _subscriber: Arc<Subscription<StringMsg>>,
    data: Arc<Mutex<Option<StringMsg>>>,
}
```
Instead of a Publisher, there is a Subscription object in the Subscriber node. The data needs to be an `Arc<Mutex<Option<StringMsg>>>` because there can be errors in the data transfer process and this can be caught by including the value of the incoming subscription in an optional.  

> #### This code defines a function named new that likely creates an instance of some SimpleSubscriptionNode.

Reviewer 3: ```suggestion
#### impl SimpleSubscriptionNode
This code defines methods for the `SimpleSubscriptionNode` `struct`. 

##### new
The `new` method creates a ROS 2 node, subscriber, and a storage location for received messages, storing them in the `struct`. 
```

Author: It has already been incorporated. What's the problem?

Reviewer 2: I don't know, what's the problem? I committed this change along with the other suggestions that @maspe36 made

        * `create_subscription`: This is creates a subscription to a specific topic.  
        * `<StringMsg, _>`: This specifies the type of message the subscription is interested in (`StringMsg`) and a placeholder (`_`) for the callback [`closure`](https://doc.rust-lang.org/book/ch13-01-closures.html) type.  
            `"publish_hello"`: This is the name of the ROS topic this node wants to subscribe to. Messages of type `StringMsg` are expected on this topic.  
        * `move |msg: StringMsg| { ... }`: This is a closure ([anonymous function](https://en.wikipedia.org/wiki/Anonymous_function)) that will be called whenever a new message arrives on the subscribed topic.
        * `msg: StringMsg`: This parameter receives the received message of type `StringMsg`. The closure body (`{...}`) uses the `Mutex` to access and update the shared data (`data_mut`) with the received message.  
3. Cloning the Shared Data:
    * `let data_mut: Arc<Mutex<Option<StringMsg>>> = Arc::clone(&data)`; This line creates another `Arc` reference (`data_mut`) pointing to the same underlying data structure as data. This allows the closure to access and modify the shared data.  

> #### this function provides a way to access and potentially use the received message data stored within the `Arc<Mutex<Option<StringMsg>>>` member variable of the `struct`. It checks if a message exists, prints it if available, or informs the user there's no message yet.

Reviewer 3: ```suggestion
##### data_callback
This function provides a way to access and potentially use the received message data stored within the `data` member variable of the `SimpleSubscriptionNode`. It checks if a message exists, prints it if available, or informs the user there's no message yet.
```

Author: done

# Writing a simple publisher and subscriber (RUST)

> * Goal: Create and run a publisher and subscriber node using Python.

Author: done

* Tutorial level: Beginner
* Time: 20 minutes
<details><summary>Background</summary>

In this tutorial you will create a pair of 
[nodes](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Nodes/Understanding-ROS2-Nodes.html) that pass information to each other via a 
[topic](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Topics/Understanding-ROS2-Topics.html) in the form of string messages. The example used here is a simple "talker" and "listener" system; one node publishes data and the other subscribes to the topic to receive that data.

Since Rust doesn't have inheritance, it's not possible to inherit from `Node` as is common practice in [`rclcpp`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html) or [`rclpy`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html).

> The code used in these examples can be found [here](https://gitlab.com/ros21923912/simple_ros2_node)  

Author: done

<div style="margin-left:20px;">
<details><summary>Side-note on dependencies</summary>

> You may be wondering why you can't just add all your ros2-specific dependencies to `cargo.toml` with `cargo add ${dependencie}` and have to edit this file manually. Here is why:

Reviewer 2: I'd avoid using shell-isms to describe terms, so something like `cargo add YOUR_DEPENDENCIES` would be nicer.

> You may be wondering why you can't just add all your ros2-specific dependencies to `cargo.toml` with `cargo add ${dependencie}` and have to edit this file manually. Here is why:

Author: done

> Almost none of the ROS2 dependencies you'll need for your ros2 rust node development currently exist on [crates.io](https://crates.io/), the main source for rust depencies. So the add command simply can't find the dependency targets. What colcon does by compiling the ros2 rust dependencies and your ros2 rust project is redirect the cargo search for dependencies directly into your `workspace/install` folder, where it'll find locally generated rust projects to use as dependencies. In particular, almost all message types will be called as dependencies for your ros2 rust project this way.

Reviewer 2: ```suggestion
Almost none of the ROS 2 dependencies you'll need for your ROS 2 Rust node development currently exist on [crates.io](https://crates.io/), the main source for rust depencies. So the add command simply can't find the dependency targets. What colcon does by compiling the ROS 2 Rust dependencies and your ROS 2 Rust project is redirect the cargo search for dependencies directly into your `workspace/install` folder, where it'll find locally generated rust projects to use as dependencies. In particular, almost all message types will be called as dependencies for your ROS 2 Rust project this way.
```

The correct spelling of ROS 2 is separately (not ROS2). It's not clear to me what you meant with ros2 rust, is it `ros2-rust` (this project) or ROS 2 Rust?

</details></div>

</details>

<details><summary>Prerequisites </summary> 

In previous tutorials, you learned how to create a [workspace](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-A-Workspace/Creating-A-Workspace.html) and [create a package](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-Your-First-ROS2-Package.html).

> A basic understanding of [RUST](https://www.rust-lang.org/) is recommended, but not entirely necessary.

Reviewer 2: ```suggestion
A basic understanding of [Rust](https://www.rust-lang.org/) is recommended, but not entirely necessary.
```

> Before developing ROS2 RUST nodes, you must follow the 

Reviewer 2: Do you mean `ros2-rust`? Or ROS 2 nodes written in Rust?

# Writing a simple publisher and subscriber (Rust)
* Goal: Create and run a publisher and subscriber node using Rust.
* Tutorial level: Beginner
* Time: 20 minutes

> <details><summary>Background</summary>

Author: In fact, that's exactly on purpose. I believe that people can follow a documentary better if they find a few interactive elements in it. In addition, this way you can only open the parts of the tutorial that you really need.

> <details><summary>Background</summary>

Reviewer: Ordinarily, I would agree with this. However, we're currently aiming to become another official client library, which means that I think we need to follow how the official docs format their files.

Perhaps someday we can write up a proposal to add in collapsible sections to the docs, but for now, I think we need to aim for consistency. Even if it's not quite optimal.

> <details><summary>Background</summary>

Author: As an alternative, I offer to rewrite the documentation into an [mdbook](https://github.com/rust-lang/mdBook). I have not yet done enough work with [rst](https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html) as it is used in the official ros2 documentation.

> <details><summary>Background</summary>

Reviewer 3: I think it's fine we aren't using rst for this documentation. I think it's more just about matching the same structure (i.e. no collapsible sections). 

Also, I think we should consider doing that formatting work in another PR. This PR has lots of great content I don't want to hold up over something like collapsible sections. 

Once all other comments are resolved I think we can merge this as is and iterate.

In this tutorial you will create a pair of 
[nodes](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Nodes/Understanding-ROS2-Nodes.html) that pass information to each other via a 
[topic](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Topics/Understanding-ROS2-Topics.html) in the form of string messages. The example used here is a simple "talker" and "listener" system; one node publishes data and the other subscribes to the topic to receive that data.

Since RUST doesn't have inheritance, it's not possible to inherit from `Node` as is common practice in [`rclcpp`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html) or [`rclpy`](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html).

> The code used in these examples can be found [here](https://github.com/Guelakais/ros2_rust/tree/better_tutorial/examples/your_package_name)  

Reviewer 2: This will be invalid once this PR is merged, can you check if relative paths work? Otherwise, replace this with `main`

Author: done

> The code used in these examples can be found [here](https://gitlab.com/ros21923912/simple_ros2_node/-/tree/more_simple_nodes?ref_type=heads)  

Author: Can you please help me recreate the build process of your Github container locally? My problem at the moment is that the build process in my installation is not throwing any errors. When I build it in your Github container it does and it takes 10 minutes every time. I find it too complicated to fight against your Github container at the moment.

</details>

<details><summary>Prerequisites </summary> 

Basic concepts of development with ROS2 should be known:
* [workspaces](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-A-Workspace/Creating-A-Workspace.html)
* [packages](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Creating-Your-First-ROS2-Package.html).

A basic understanding of [Rust](https://doc.rust-lang.org/book/) is recommended, but not entirely necessary.
Before developing [ros2-rust](https://github.com/ros2-rust/ros2_rust) nodes, you must follow the 
[installation instructions](https://github.com/ros2-rust/ros2-rust/blob/main/README.md) for [`rclrs`](https://docs.rs/rclrs/latest/rclrs/).

</details>

<details><summary>Tasks </summary> 
<div style="margin-left:20px;"><details><summary>Create a Package</summary>

Currently, building a package for ros2-rust is different 
from building packages for [Python](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html) or [C/C++](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Publisher-And-Subscriber.html).  

> First, you'll need to create and go into a standard [cargo](https://doc.rust-lang.org/cargo/) 

Author: It's done

project as follows:
```
cargo new your_package_name && cd your_package_name

> ```

Reviewer 2: @Guelakais please address this

> ```

Author: That was too inconsistent for me. I have replaced each `shell` and `console` hint with an `sh`. so it is consistent for normal terminal calls.

```xml
<package format="3">
  <name>your_package_name</name>
  <version>0.0.0</version>
  <description>TODO: Package description.</description>
  <maintainer email="user@todo.todo">user</maintainer>
  <license>TODO: License declaration.</license>

  <depend>rclrs</depend>
  <depend>std_msgs</depend>

  <export>
    <build_type>ament_cargo</build_type>
  </export>
</package>
```
By taking a look at your package, for example by typing [`tree .`](https://www.geeksforgeeks.org/tree-command-unixlinux/) inside your package, and you'll see a structure similar to the following:  
```
├── Cargo.toml
├── package.xml
└── src
    └── main.rs

2 directories, 3 files
```
Of course, you can use any capable editor or even your file explorer to do this.  

</details>

<details><summary>Write the publisher node</summary>

To construct a node, replace the code in your `main.rs` file with the [following](https://gitlab.com/ros21923912/simple_ros2_node/-/blob/more_simple_nodes/src/simple_publisher.rs?ref_type=heads):  

> ```

Author: done

* `let mut iterator: i32 = 0;`: Initializes a counter variable for message content.
* `thread::spawn(move || -> () { ... });`: Spawns a new thread with a closure: `iter::repeat(()).for_each(|()| { ... });`: Creates an infinite loop using `iter::repeat`.

4. Publishing Loop within Thread:

* `thread::sleep(Duration::from_millis(1000));`: Pauses the thread for 1 second (1 Hz publishing rate).
* `iterator = publisher_other_thread.publish_data(iterator).unwrap();`: Calls the publish_data method on the publisher_other_thread to publish a message with the current counter value. Increments the iterator for the next message.

5. Main Thread Spin:
* `rclrs::spin(publisher.node.clone());`: Keeps the main thread running, processing ROS 2 events and messages. Uses a cloned reference to the node to ensure it remains active even with other threads.

</details>
</details>
<details><summary>Having several ROS2 Rust nodes in one Package</summary>

Of course, you can write for each node you want to implement its own package, and that can have it's advantages. I implore you to use some cargo tricks and add some binary targets to your `cargo.toml`. That could look like this:

> ```

Author: done
